unpacker.go

```diff
@@ -87,85 +88,128 @@ func (u *unpacker) unpack(ctx context.Context, config ocispec.Descriptor, layers
 		a  = u.c.DiffService()
 		cs = u.c.ContentStore()
 
-		states []layerState
-		chain  []digest.Digest
+		chain []digest.Digest
+
+		fetchOffset int
+		fetchC      []chan struct{}
+		fetchErr    chan error
 	)
+
+	// If there is an early return, ensure any ongoing
+	// fetches get their context cancelled
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+EachLayer:
 	for i, desc := range layers {
-		states = append(states, layerState{
-			layer: rootfs.Layer{
-				Blob: desc,
-				Diff: ocispec.Descriptor{
-					MediaType: ocispec.MediaTypeImageLayer,
-					Digest:    diffIDs[i],
-				},
-			},
-		})
-	}
-	for {
-		var layer ocispec.Descriptor
-		select {
-		case layer = <-u.updateCh:
-		case <-ctx.Done():
-			return ctx.Err()
+		parent := identity.ChainID(chain)
+		chain = append(chain, diffIDs[i])
+
+		chainID := identity.ChainID(chain).String()
+		if _, err := sn.Stat(ctx, chainID); err == nil {
+			// no need to handle
+			continue
+		} else if !errdefs.IsNotFound(err) {
+			return errors.Wrapf(err, "failed to stat snapshot %s", chainID)
 		}
-		log.G(ctx).WithField("desc", layer).Debug("layer downloaded")
-		for i := range states {
-			if states[i].layer.Blob.Digest != layer.Digest {
-				continue
-			}
-			// Different layers may have the same digest. When that
-			// happens, we should continue marking the next layer
-			// as downloaded.
-			if states[i].downloaded {
-				continue
+
+		labelOpt := snapshots.WithLabels(map[string]string{
+			"containerd.io/snapshot.ref": chainID,
+		})
```

Comment:
Could you pass the layer's basic information?
- Reference name of the image which the layer belongs to
- Digest(not ChainID, the one written in the OCI manifest) of the layer

CRFS needs them to get the layer URL.

https://github.com/ktock/remote-snapshotter/blob/master/filesystems/stargz/fs.go#L152

CVMFS also seems to use the digest to get the path on the filesystem.

https://github.com/ktock/remote-snapshotter/blob/bba94869bfde513a6e7f9d3a009b08f21102b263/filesystems/cvmfs/fs.go#L44

Reply:
I think we can add an option to do it, it doesn't make sense to do by default as this is very particular to one possible backend implementation. Generally speaking this information does not identify the snapshot and for any given snapshot it could end up having multiple values. Either way, out of scope for this PR.

Reply:
So for some of these type of implementations, I would favor such a snapshot actually making use of the content store interface for accessing the data. We have already defined a label format there for referencing the source. Likewise referring to a piece of content from a snapshot is an acceptable pattern, although in this case the reference would need to be passed down to the backend.

Reply:
Thank you for your idea!

So we seem to need another patch for storing data X(might be chainID string?) which is label by distribution source information(`containerd.io/distribution.source`) to the content store.

There's room for consideration about what should X be, but I think the X being chainID string(+some dedicated prefix) is good because snapshotter can easily refer it without any additional labels through `Prepare`.